clippy_utils/src/hir_utils.rs

```diff
@@ -29,6 +30,30 @@ pub struct SpanlessEq<'a, 'tcx> {
     maybe_typeck_results: Option<&'tcx TypeckResults<'tcx>>,
     allow_side_effects: bool,
     expr_fallback: Option<Box<dyn FnMut(&Expr<'_>, &Expr<'_>) -> bool + 'a>>,
+    /// This adds an additional type comparison to locals that insures that even the
+    /// inferred of the value is the same.
+    ///
+    /// **Example**
+    /// * Context 1
+    /// ```ignore
+    /// let vec = Vec::new();
+    /// vec.push("A string");
+    /// ```
+    ///
+    /// * Context 2
+    /// ```ignore
+    /// let vec = Vec::new();
+    /// vec.push(0); // An integer
+    /// ```
+    ///
+    /// Only comparing the first local definition would usually return that they are
+    /// equal, since they are identical. However, they are different due to the context
+    /// as they have different inferred types.
+    ///
+    /// This option enables or disables the specific check of the inferred type.
+    ///
+    /// Note: This check will only be done if `self.maybe_typeck_results` is `Some()`.
+    check_inferred_local_types: bool,
```

Comment:
Any reason to not simply always do this?

Reply:
This lint is currently the only chase where equality is evaluated on a per statement level, the other use cases inspect the entire context and catch such cases without the need of this additional check. So, we could check this every time, but it's often unnecessary and might alter the result of some evaluations as `let vec = Vec::new()` == `let vec = Vec::new()` would not always be the same anymore.

Reply:
I think it is not unlikely that this will fix other false positives. So I think its most safe (and simple) to just always have the check.

Reply:
Fair point, I've removed the configuration and all test still pass :)